lib/base.js

```diff
@@ -0,0 +1,790 @@
+  },
+
+  /*
+   * Some attributes may be elements. This is domain-specific knowledge, so
+   * a subclass *MUST* define the attribute element names to convert. This
+   * method handles the actual conversion when loading.
+   */
+  convertAttributesToElements: function(conversionFunc) {
+    for (var i = 0; i < this._attributeElementKeys.length; i++) {
+      var key = this._attributeElementKeys[i];
+
+      if (this.attributes[key]) {
+        this.attributes[key] = conversionFunc(this.attributes[key]);
+      }
+
+    }
+  },
+
+  fromRefract: function(dom) {
+    this.meta = new Meta(dom.meta);
+    this.attributes = dom.attributes;
+    this.content = dom.content;
+
+    this.convertAttributesToElements(function(attribute) {
+      return registry.fromRefract(attribute);
+    });
+
+    if (this.element !== dom.element) {
+      this.element = dom.element;
+    }
+
+    return this;
+  },
+
+  fromCompactRefract: function(tuple) {
+    this.meta = new Meta(tuple[1]);
+    this.attributes = tuple[2];
+    this.content = tuple[3];
+
+    this.convertAttributesToElements(function(attribute) {
+      registry.fromCompactRefract(attribute);
+    });
+
+    if (this.element !== tuple[0]) {
+      this.element = tuple[0];
+    }
+
+    return this;
+  },
+
+  set: function(content) {
+    this.content = content;
+    return this;
+  },
+
+  equals: function(value) {
+    return _.isEqual(this.toValue(), value);
+  }
+});
+
+Object.defineProperty(ElementType.prototype, 'element', {
+  get: function() {
+    // Returns 'element' so we don't have undefined as element
+    return this._storedElement || 'element';
+  },
+  set: function(element) {
+    this._storedElement = element;
+  }
+});
+
+function NullType() {
+  ElementType.apply(this, arguments);
+  this.element = 'null';
+}
+
+NullType.prototype = _.create(ElementType.prototype, {
+  primitive: function() {
+    return 'null';
+  },
+
+  set: function() {
+    return new Error('Cannot set the value of null');
+  }
+});
+
+function StringType() {
+  ElementType.apply(this, arguments);
+  this.element = 'string';
+}
+
+StringType.prototype = _.create(ElementType.prototype, {
+  primitive: function() {
+    return 'string';
+  }
+});
+
+// TODO: Add tests
+Object.defineProperty(StringType.prototype, 'length', {
+  get: function() {
+    return this.content.length;
+  }
+});
+
+function NumberType() {
+  ElementType.apply(this, arguments);
+  this.element = 'number';
+}
+
+NumberType.prototype = _.create(ElementType.prototype, {
+  primitive: function() {
+    return 'number';
+  }
+});
+
+function BooleanType() {
+  ElementType.apply(this, arguments);
+  this.element = 'boolean';
+}
+
+BooleanType.prototype = _.create(ElementType.prototype, {
+  primitive: function() {
+    return 'boolean';
+  }
+});
+
+function Collection() {
+  ElementType.apply(this, arguments);
+}
+
+// TODO: Iterators for ES5
+// get [Symbol.iterator]() {
+//   return this.content[Symbol.iterator];
+// }
```

Comment:
I guess we'll need to check if `Symbol` is even available, then attach this when it's possible. Probably need to play with this a bit.